### ISSUE TITLE
Thread safety analysis and temporaries

### ISSUE THREAD

**Issue by leonid-s-usov** [llvm/llvm-project#98004]
Hi!

It appears that when temporaries are involved, the analyzer ignores the destructors, which causes the static analysis warnings. Please see this compiler explorer example: https://godbolt.org/z/r4fa5Kd7r

I am trying to create a movable scoped lockable, but the issue is preventing me from being able to generically return and move-construct the scoped objects.

There are two issues, actually. One is the temporary object whose destructor isn't considered, and other is a simple wrapper struct, where the member in the struct is scoped lockable. Both cases result in a warning.

```
    std::cout << "=== temporary destructor ignored" << std::endl;
    {
        ValueReader reader(&value);

        auto val = [reader2 = ValueReader(std::move(reader))]() mutable {
            auto reader3 = std::move(reader2);
            std::cout << "in a lambda 2" << std::endl;
            return reader3.value();
        }();
    }
```

```
    std::cout << "=== member destructor ignored" << std::endl;

    {
        ValueReader reader(&value);

        struct Wrapper {
            ValueReader wrapped_reader;
        };

        auto wr = Wrapper { std::move(reader) };

        assert(wr.wrapped_reader.value() == value.value);
    }
```

**Comment by usx95** [comment]
CC @aaronpuchert @delesley

**Comment by leonid-s-usov** [comment]
I suspect now that the case with the lambda capture initialization is directly related to the struct member issue, plus the mishandling of the `CXXBindTemporaryExpr` which registers a call to constructor but loses track of the object afterwards

**Comment by usx95** [comment]
The problem is specific to how we handle member destructors.

https://godbolt.org/z/34xzeozGn
```cpp
int main() {
    ProtectedValue value = {10, Mutex{}};
    struct Wrapper {
        ValueReader r;
    };
    Wrapper wr = Wrapper{ValueReader(&value)};
}
```

<details>
<summary>CFG</summary>

```
 [B1]
   1: 10
   2: /*implicit*/(_Bool)0
   3: {}
   4: Mutex[B1.3] (CXXFunctionalCastExpr, NoOp, Mutex)
   5: {[B1.1], [B1.4]}
   6: ProtectedValue value = {10, Mutex{}};
   7: value
   8: &[B1.7]
   9: [B1.8] (CXXConstructExpr, ValueReader)
  10: [B1.9] (BindTemporary)
  11: ValueReader([B1.10]) (CXXFunctionalCastExpr, ConstructorConversion, ValueReader)
  12: {[B1.11]}
  13: [B1.12] (BindTemporary)
  14: Wrapper[B1.13] (CXXFunctionalCastExpr, NoOp, Wrapper)
  15: Wrapper wr = Wrapper{ValueReader(&value)};
  16: [B1.15].~Wrapper() (Implicit destructor)
   Preds (1): B2
   Succs (1): B0
```
</details>

I think what is happening here is that the dtor of the temporary `~ValueReader` is called by `Wrapper::~Wrapper()` and not inside `main()`. This is expected.

But this makes the analysis see one lock (`@ValueReader::ValueReader(ProtectedValue*)`  but only **no** unlock.

```llvm
entry:
  %value = alloca %struct.ProtectedValue, align 4
  %wr = alloca %struct.Wrapper, align 8
  call void @llvm.memcpy.p0.p0.i64(ptr align 4 %value, ptr align 4 @__const.main.value, i64 8, i1 false)
  %r = getelementptr inbounds %struct.Wrapper, ptr %wr, i32 0, i32 0
  call void @ValueReader::ValueReader(ProtectedValue*)(ptr noundef nonnull align 8 dereferenceable(8) %r, ptr noundef %value)
  call void @main::Wrapper::~Wrapper()(ptr noundef nonnull align 8 dereferenceable(8) %wr) #6
  ret i32 0
```

We seem to [ignore member destructors](https://github.com/llvm/llvm-project/blob/3670e7f86cc69692432e57828a0acf1a5abd3156/clang/lib/Analysis/ThreadSafety.cpp#L2464) while handling destructor calls.

**Comment by leonid-s-usov** [comment]
> We seem to [ignore member destructors](https://github.com/llvm/llvm-project/blob/3670e7f86cc69692432e57828a0acf1a5abd3156/clang/lib/Analysis/ThreadSafety.cpp#L2464) while handling destructor calls.

Correct. Though the second piece of the puzzle is that `bind temporary` step which calls the constructor. It's going to be tricky, since the member constructor isn't called explicitly, but "inherited" from that temporary via binding

**Comment by leonid-s-usov** [comment]
> The problem is specific to how we handle member destructors.

I agree that the lambda capture initialization is related to struct members. However, the [third case](https://godbolt.org/z/Wq8fPrfTj) has no members, but it has `CXXBindTemporaryExpr` which constructs the return value directly into the `reader3` var.

```
    {
        ValueReader reader(&value);

        auto reader3 = [&]() -> ValueReader {
            auto reader2 = std::move(reader);
            std::cout << "in a lambda 3" << std::endl;
            return std::move(reader2);
        }();
    }
```

**Comment by aaronpuchert** [comment]
> I am trying to create a movable scoped lockable.

The problem with moving is not so much support for moving itself, but the accompanying rewrite of the managed capabilities. On construction of a scoped capability, we store expressions referencing the managed capabilities (the mutexes managed by the scoped object), obtained by rewriting the attribute expression in the context of construction. At this point, by definition, we can still reference the mutex, but after a move, this might no longer be the case.
```c++
struct Foo {
  Mutex mu;
};
struct __attribute__((scoped_lockable)) Locker {
  Locker(Foo& foo) __attribute__((exclusive_lock_function(foo.mu)));
};
void bar(Foo* ptr) {
  Locker lock(*ptr);
}
```
In this case, we substitute `*ptr` for the formal parameter `foo` and get `(*ptr).mu` as the capability managed by `lock`.

See below for an example where it doesn't work, because we have a scoped lockable that's not a local variable.

> It kind of works, but there are some cases that generate false positives. It seems that there are two interconnected problems:
> 
> 1. The lack of support for `scoped_lockable` members

This is probably the most interesting case for moving, but it breaks the connection between scoped capability and managed capability. If we move the scoped capability into a different context, the managed capability might no longer be "expressible":
```c++
struct Foo {
  Mutex mu;
  Wrapper lock();  // Wrapper from original example
};

void bar() {
  Foo foo;
  Wrapper wrapper = foo.lock();
  // How does wrapper.wrapped_locker reference foo.mu?
}
```
This is a problem of scopes and lookup: there is no way to reference `foo.mu` from inside the wrapper. The wrapper cannot refer to the context it is contained in, because it doesn't know anything about that context. This is not an issue if you return the `MutexLocker` itself, but as a member of another `struct` it loses the connection.

> 2. Some issue with temporaries, presumably around copy elision via temporary binding

Maybe you can help me spot the copy elision in your examples, but fundamentally you're right: in case of mandatory copy elision, we don't have a constructor call at all, so attributes on a constructor cannot be taken into account. But I'm not sure what kind of problems this generates. Note that I changed the behavior in case of copy elision in https://reviews.llvm.org/D129755 (54bfd04846156dbd5e0a6b88f539c3d4569a455f). We take the attributes on the function that returns the scoped lockable object in that case.

> Here is a compiler explorer example with all of the observed issues: https://godbolt.org/z/vfcb78jqc
> In case the above compiler explorer snippet is not available, please see the full source code under the cut:
> 
> ```c++
>     // we mark this as exclusive lock because
>     // although we're taking the mutex locked already
>     // the static analysis doesn't know that other.pmutex
>     // and this->pmutex are referencing the same object
>     // With the current accounting we should pretend that
>     // we took our new lock, and the other will unlock
>     // in the destructor - not in the runtime, but static analysis-wise.
>     MutexLocker(MutexLocker &&other) __attribute__((exclusive_lock_function(*pmutex))) { /*...*/ }
> ```

Conceptually this doesn't look right. The move constructor doesn't acquire anything, and it should probably not need an annotation at all.

> ```c++
>     std::cout << "=== no warnings in most cases" << std::endl;
>     {
>         MutexLocker l(&m);
> 
>         MutexLocker l2 = std::move(l);
> 
>         auto val = [&]() {
>             auto l3 = std::move(l2);
>             std::cout << "in lambda 1" << std::endl;
>             return 10;
>         }();
>     }
> ```

The first part should be easy to support, but uninteresting. The second part is more problematic. The lambda captures the scoped lockable, but probably not the mutex itself. So how do we express the relationship of the lockable to the mutex, if it is unknown to the lambda? If it is known, it would be known [via alias](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-alias-analysis).

Also, if we don't see the mutex anymore, what about data that has `guarded_by` attribute? That likely lives in the same scope.

> The problem is specific to how we handle member destructors.

So far we have restrained from looking inside called functions, but since destructors are known to destruct all members we could probably take that into account. However, destruction is not the biggest issue here, it's the rewriting of expressions.
